ssh2/session.pyx

```diff
@@ -45,18 +50,49 @@ LIBSSH2_HOSTKEY_TYPE_UNKNOWN = c_ssh2.LIBSSH2_HOSTKEY_TYPE_UNKNOWN
 LIBSSH2_HOSTKEY_TYPE_RSA = c_ssh2.LIBSSH2_HOSTKEY_TYPE_RSA
 LIBSSH2_HOSTKEY_TYPE_DSS = c_ssh2.LIBSSH2_HOSTKEY_TYPE_DSS
 
+cdef ssize_t _send_callback(int fd,
+                            char* buffer,
+                            size_t length,
+                            int flags,
+                            void** abstract) with gil:
+    global errno
+    try:
+        bp = PyBytes_FromStringAndSize(buffer, length)
+        res = (<Session>deref(abstract))._send_callback(fd, bp, flags)
+        return length
+    except Exception as e:
```

Comment:
> It is up to caller to handle exceptions in callback function which cannot happen if they are not raised.

If it is up to the caller to handle exceptions, why do we need the  `except -1`? Am I okay to just remove the `try..catch` and `errno` logic entirely?

Reply:
`except -1` is needed because without it Cython will not propagate the exception to python land (Py_SetError will not be called). Meaning the exception will be hidden.

Yes, removing the try/catch sounds good - can leave it up to caller to handle.